prow/spyglass/gcsartifact.go

```diff
@@ -0,0 +1,185 @@
+}
+
+// NewGCSArtifact returns a new GCSArtifact with a given handle, canonical link, and path within the job
+func NewGCSArtifact(ctx context.Context, handle *storage.ObjectHandle, link string, path string, sizeLimit int64) *GCSArtifact {
+	return &GCSArtifact{
+		handle:    handle,
+		link:      link,
+		path:      path,
+		sizeLimit: sizeLimit,
+		ctx:       ctx,
+	}
+}
+
+func fieldsFor(a *GCSArtifact) logrus.Fields {
+	return logrus.Fields{
+		"artifact": a.path,
+	}
+}
+
+// Size returns the size of the artifact in GCS
+func (a *GCSArtifact) Size() (int64, error) {
+	attrs, err := a.handle.Attrs(a.ctx)
+	if err != nil {
+		return 0, err
+	}
+	return attrs.Size, nil
+}
+
+// JobPath gets the GCS path of the artifact within the current job
+func (a *GCSArtifact) JobPath() string {
+	return a.path
+}
+
+// CanonicalLink gets the GCS web address of the artifact
+func (a *GCSArtifact) CanonicalLink() string {
+	return a.link
+}
+
+// ReadAt reads len(p) bytes from a file in GCS at offset off
+func (a *GCSArtifact) ReadAt(p []byte, off int64) (n int, err error) {
+	gzipped, err := a.gzipped()
+	if err != nil {
+		return 0, err
+	}
+	if gzipped {
+		return 0, viewers.ErrGzipOffsetRead
+	}
+	reader, err := a.handle.NewRangeReader(a.ctx, off, int64(len(p)))
+	defer reader.Close()
+	if err != nil {
+		return 0, err
+	}
+	n, err = reader.Read(p)
+	return n, err
+}
+
+// ReadAtMost reads at most n bytes from a file in GCS. If the file is compressed (gzip) in GCS, n bytes
+// of gzipped content will be downloaded and decompressed into potentially GREATER than n bytes of content.
+func (a *GCSArtifact) ReadAtMost(n int64) ([]byte, error) {
+	reader, err := a.handle.NewRangeReader(a.ctx, 0, n+1)
+	defer reader.Close()
+	if err != nil {
+		return []byte{}, err
+	}
+	var p []byte
+	var e error
+	gzipped, err := a.gzipped()
+	if err != nil {
+		return []byte{}, err
+	}
+	if gzipped {
+		gReader, err := gzip.NewReader(reader)
+		if err != nil {
+			return []byte{}, err
+		}
+		p, e = ioutil.ReadAll(gReader)
+	} else {
+		p, e = ioutil.ReadAll(reader)
+	}
+	return p, e
+}
+
+// ReadAll will either read the entire file or throw an error if file size is too big
+func (a *GCSArtifact) ReadAll() ([]byte, error) {
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, err
+	}
+	if size > a.sizeLimit {
+		return []byte{}, viewers.ErrFileTooLarge
+	}
+	reader, err := a.handle.NewReader(a.ctx)
+	defer reader.Close()
+	if err != nil {
+		return []byte{}, err
+	}
+	p, err := ioutil.ReadAll(reader)
+	return p, err
+}
+
+// ReadTail reads the last n bytes from a file in GCS
+func (a *GCSArtifact) ReadTail(n int64) ([]byte, error) {
+	gzipped, err := a.gzipped()
+	if err != nil {
+		return []byte{}, err
+	}
+	if gzipped {
+		return []byte{}, viewers.ErrGzipOffsetRead
+	}
+	size, err := a.Size()
+	if err != nil {
+		return []byte{}, err
+	}
+	offset := size - n
+	reader, err := a.handle.NewRangeReader(a.ctx, offset, -1)
+	defer reader.Close()
+	if err != nil {
```

Comment:
Personally I think you should own your exceptions.

So this should be something like:
`err := fmt.Errorf("failed range read: %v", err)`

rather than just returning someone else's error


The key thing that the `ReadTail()` caller needs to distinguish that they cannot do own their own is whether the err came from:
```
a.gzipped()
a.Size()
a.handle.NewRangeReader()
ioutil.ReadAll()
```

So you ought to encode this info in the error you return